Forecasting concepts session (#126)
nfidd/nfidd: sessions/forecasting-concepts.qmd

library("here")
library("cmdstanr")
library("tidybayes")
library("scoringutils")
```

::: {.callout-tip}
The best way to interact with the material is via the [Visual Editor](https://docs.posit.co/ide/user/ide/guide/documents/visual-editor.html) of RStudio.
If not using the Visual Editor please remember that the code in the session needs to be run inside the course repository so that the `here()` commands below find the stan model files.
:::

# Introduction to forecasting as an epidemiological problem, and its relationship with nowcasting and R estimation

See the lecture slides.

> 

Author: Can you do this? I'm not sure I follow what you mean enough to write an issue.

> 

Reviewer: yes

##### Interpretation

- **First Term ($ \mathbb{E}_{X \sim D}[|X - y|] $)**: This term penalizes the forecast based on how far the predicted values are from the observed value. It ensures that the forecast is accurate in terms of proximity to the actual observation.

- **Second Term ($ \frac{1}{2} \mathbb{E}_{X, X' \sim D}[|X - X'|] $)**: This term accounts for the spread of the forecasted distribution. It penalizes forecasts that are too uncertain or have a wide spread. By subtracting this term, the CRPS rewards forecasts that are not only accurate but also confident (i.e., have a narrow spread).
:::

Whilst the CRPS is a very useful metric it can be difficult to interpret in isolation. It is often useful to compare the CRPS of different models or to compare the CRPS of the same model under different conditions. For example, lets compare the CRPS across different forecast horizons.

```{r}
sc_scores |>
  summarise_scores(by = "horizon") |>
  ggplot(aes(x = horizon, y = crps)) +
  geom_point()
```

> and across different forecast horizons.

Reviewer: ```suggestion
and at different time points.
```

  - The CRPS scores increase with forecast horizon.
  - As the CRPS is an absolute measure it is hard to immediately know if the CRPS increasing with forecast date indicates that the model is performing worse.
  - However, the CRPS increasing with forecast horizon is a sign that the model is struggling to capture the longer term dynamics of the epidemic.
:::

### PIT histograms

As well as the CRPS we can also look at the calibration and bias of the model. Calibration is the agreement between the forecast probabilities and the observed frequencies. Bias is a measure of how likely the model is to over or under predict the observed values.

There are many ways to assess calibration and bias but one common way is to use a probability integral transform (PIT) histogram. This is a histogram of the cumulative distribution of function of a forecast evaluated at the observed value.

::: {.callout-tip}
## Interpreting the PIT histogram
- Ideally PIT histograms should be uniform. 
- If is a U shape then the model is overconfident and if it is an inverted U shape then the model is underconfident. 
- If it is skewed then the model is biased towards the direction of the skew.
:::

> ::: {.callout-tip, collapse="true"}

Reviewer: ```suggestion
::: {.callout-tip collapse="true"}
```

$$ U = F_X(k) + v \cdot (F_X(k) - F_X(k-1)) $$

where:
- $ k $ is the observed integer value.
- $ F_X(k) $ is the CDF evaluated at $ k $.
- $ v $ is a random variable uniformly distributed on $[0, 1]$.

This transformation ensures that $ U $ is uniformly distributed on $[0, 1]$ if the forecasted distribution $ F_X $ is correctly specified.
:::

Let's first look at the overall PIT histogram.

```{r pit-histogram}
 sc_forecasts |>
  get_pit(by = "model") |>
  plot_pit()
```

As before lets look at the PIT histogram by forecast horizon (to save space we will group horizons)

```{r pit-histogram-horizon}
sc_forecasts |> 
  mutate(group_horizon = case_when(
    horizon <= 3 ~ "1-3",
    horizon <= 7 ~ "4-7",
    horizon <= 14 ~ "8-14"
  )) |>
  get_pit(by = "group_horizon") |>
  plot_pit() +
  facet_wrap(~group_horizon)
```

and then for different forecast dates.

```{r pit-histogram-date}
 sc_forecasts |>
  get_pit(by = "target_day") |>
  plot_pit() +
  facet_wrap(~target_day)
```

::: {.callout-tip}
## Take 5 minutes
What do you think of the PIT histograms?
Do they look well calibrated?
Do they look biased?
:::

::: {.callout-note collapse="true"}
## Solution

> - It looks like the model is biased towards over predicting and that this bias gets worse at longer forecast horizons.

Reviewer: ```suggestion
- It looks like the model is biased towards overpredicting and that this bias gets worse at longer forecast horizons.
```

> - Looking over forecast dates it looks like much of  this bias is coming from near the outbreak peak where the model is consistently over predicting but the model is also over predicting at other times.

Reviewer: ```suggestion
- Looking over forecast dates it looks like much of  this bias is coming from near the outbreak peak where the model is consistently overpredicting but the model is also over predicting at other times.
```

:::

## Scoring on the log scale

> We can also score on the log scale.

Reviewer: ```suggestion
We can also score on the logarithmic scale.
```

> This can be useful if we are interested in the relative performance of the model at different scales of the data.
> For example, if we are interested in the model's performance at capturing the exponential growth phase of the epidemic.

Reviewer: ```suggestion
This can be useful if we are interested in the relative performance of the model at different scales of the data, for example if we are interested in the model's performance at capturing the exponential growth phase of the epidemic.
```

```{r log-pit-histogram-date}
log_sc_forecasts |>
  get_pit(by = "target_day") |>
  plot_pit() +
  facet_wrap(~target_day)
```

::: {.callout-tip}
## Take 5 minutes
What do you think of the PIT histograms?
Do they look well calibrated?
Do they look biased?
:::

::: {.callout-note collapse="true"}
## Solution
- The overall PIT histograms suggest that the model is less biased to over predict on the log scale than the natural scale but it is still biased. This makes sense when we think back to the comparison of reproduction number estimates and forecasts we made earlier where the model was consistently over predicting on the reproduction number scale.
- By forecast horizon the model is still biased towards over predicting but this bias is less pronounced than on the natural scale.
- Towards the end and beginning of the forecast period the model appears to be well calibrated on the log scale but is biased towards over predicting in the middle of the forecast period.
- This matches with our knowledge of the underlying reproduction number which were initially constant and then began to decrease only to stabilise towards the end of  the outbreak. 
:::

# Going further

> - What other ways could we summarise the performance of the forecasts?

Reviewer: ```suggestion
- In which other ways could we summarise the performance of the forecasts?
```

- What other metrics could we use?

> - There is no one-size-fits-all approach to forecast evaluation, often you will need to use a combination of metrics to understand the performance of your model  and typically the metrics you use will depend on the context of the forecast. What attributes of the forecast are most important to you?

Reviewer: ```suggestion
- There is no one-size-fits-all approach to forecast evaluation, often you will need to use a combination of metrics to understand the performance of your model and typically the metrics you use will depend on the context of the forecast. What attributes of the forecast are most important to you?
```

Epidemiological forecasting is closely related to nowcasting and, when using mechanistic approaches, estimation of the reproduction number. In fact, the model we will use for forecasting is the same as the model we used for nowcasting and estimation of the reproduction number. The only difference is that we will extend the model into the future and for the purpose of this session we will remove the nowcasting part of the model.

# Extending a model into the future

The model we introduced in the [renewal session]() the reproduction number using a random walk, then used a discrete renewal process to model the number of infections, and convolved these with a delay distribution to model the number of onsets with Poisson observation error.
Based on what we found in the [nowcasting session]() this seems like a reasonable model for the data and so we might want to use it to forecast into the future.
We can do this by extending the model into the future by simulating the model forward in time. 
This is known as a generative model.
We actually already introduced a model that could do this in the renewal session (remember `h` from the data list in the renewal session?).

> So what does this look like in code? Lets load in the model again and take a look.

Reviewer 2: IMO, I think we should introduce the forecasting piece here and not in the renewal equation session.
